optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/tri/TriAnchor.java

```diff
@@ -0,0 +1,142 @@
+                on(scoreHolderGlobal, getAMetadata().getVariableDeclaration(), getBMetadata().getVariableDeclaration())
+                        .execute((drools, scoreHolder, __, ___) -> {
+                            RuleContext kcontext = (RuleContext) drools;
+                            scoreHolder.impactScore(kcontext);
+                        });
+        return Arrays.asList(getAMetadata().getPattern(), getBMetadata().getPattern(), getCMetadata().getPattern(),
+                consequence);
+    }
+
+    public List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal,
+            ToIntTriFunction matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a, b, c) -> {
+            int weightMultiplier = matchWeighter.applyAsInt(inline(a), inline(b), inline(c));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    public List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal,
+            ToLongTriFunction matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a, b, c) -> {
+            long weightMultiplier = matchWeighter.applyAsLong(inline(a), inline(b), inline(c));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    public <A, B, C> List<RuleItemBuilder<?>> terminateWithScoring(
+            Global<? extends AbstractScoreHolder> scoreHolderGlobal, TriFunction<A, B, C, BigDecimal> matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a, b, c) -> {
+            BigDecimal weightMultiplier = matchWeighter.apply(inline(a), inline(b), inline(c));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    private <A, B, C, ScoreHolder extends AbstractScoreHolder> List<RuleItemBuilder<?>> terminateWithScoring(
+            Global<ScoreHolder> scoreHolderGlobal, Block5<Drools, ScoreHolder, A, B, C> consequenceImpl) {
+        ConsequenceBuilder._4<ScoreHolder, A, B, C> consequence =
+                on(scoreHolderGlobal, (Declaration<A>) getAMetadata().getVariableDeclaration(),
+                        (Declaration<B>) getBMetadata().getVariableDeclaration(),
+                        (Declaration<C>) getCMetadata().getVariableDeclaration())
+                        .execute(consequenceImpl);
+        return Arrays.asList(getAMetadata().getPattern(), getBMetadata().getPattern(), getCMetadata().getPattern(),
+                consequence);
+    }
+
+    private static <A> A inline(Object item) {
+        if (item instanceof LogicalTuple) {
+            return ((LogicalTuple) item).getItem(0);
+        }
```

Comment:
This isn't inlining.
The instanceof check happens for every match.

An inlining approach, would return a Supplier lamdba that returns the identity() if the metadata says it's a LocalTuple (note that there are no item instances yet at that time) and the Logicaltuple getItem() extraction lamdba otherwise.

Reply:
Yeah, that will be easy to fix.